lightningd/pay.c

```diff
@@ -487,26 +487,26 @@ remote_routing_failure(const tal_t *ctx,
 	return routing_failure;
 }
 
-void payment_store(struct lightningd *ld,
-		   const struct sha256 *payment_hash, u64 partid)
+void payment_store(struct lightningd *ld, struct wallet_payment *payment TAKES)
 {
 	struct sendpay_command *pc;
 	struct sendpay_command *next;
-	const struct wallet_payment *payment;
+	/* Need to remember here otherwise wallet_payment_store will free us. */
+	bool ptaken = taken(payment);
```

Comment:
Would `if (taken(payment)) tal_steal(tmpctx, payment);` work as well? `/* keep payment alive across the loop, but schedule it for freeing later. */`

Reply:
Interesting idea, I prefer freeing as soon as possible, so we don't have lingering things, and `valgrind` can get to work, and point out potential use-after-free issues that are masked by us deferring free to the `io_loop`. But I'm sure it'd work perfectly fine to steal onto `tmpctx`.